internal/controller/customreplicaset_controller.go

```diff
@@ -74,13 +80,204 @@ func (r *CustomReplicaSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
+	t := time.Now()
+	newRevision := v1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-controller-revision-%d", cr.Name, t.UnixNano()),
+			Namespace: cr.Namespace,
+			Labels: map[string]string{
+				"owner": cr.Name,
+			},
+		},
+		Data:     revisionData,
+		Revision: latestRevisionNumber + 1,
+	}
+
+	// Delete oldest revision if over revision limit
+	if len(controllerRevisions.Items) >= int(cr.Spec.RevisionHistoryLimit) {
+		oldestRevision, err := r.getControllerRevisionAtIndex(ctx, cr, controllerRevisions, 0)
+		if err != nil {
+			return err
+		}
+
+		if err := r.Delete(ctx, oldestRevision); err != nil {
+			return err
+		}
+	}
+
+	// Create Revision
+	if err := r.Create(ctx, &newRevision); err != nil {
+		return err
+	}
+	// Set CR as its owner
+	if err := controllerutil.SetControllerReference(cr, &newRevision, r.Scheme); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Update 'revisionNumber' label of the CustomReplicaSet
+func (r *CustomReplicaSetReconciler) updateCRRevisionNumLabel(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, revisionNumber int64) error {
+	if cr.Labels == nil {
+		cr.Labels = map[string]string{}
+	}
+	cr.Labels["revisionNumber"] = strconv.FormatInt(revisionNumber, 10)
+	if err := r.Update(ctx, cr); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func hashControllerRevisionData(revisionData runtime.RawExtension) (string, error) {
+	jsonData, err := json.Marshal(revisionData)
+	if err != nil {
+		return "", err
+	}
+
+	// Compute SHA256 hash of PodTemplate
+	hash := sha256.Sum256(jsonData)
+
+	// Convert the hash to hexadecimal string
+	hashString := fmt.Sprintf("%x", hash)
+
+	return hashString, nil
+}
+
+func (r *CustomReplicaSetReconciler) getAllControllerRevisions(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, sorted bool) (*v1.ControllerRevisionList, error) {
+	controllerRevisions := &v1.ControllerRevisionList{}
+	if err := r.List(ctx, controllerRevisions, client.InNamespace(cr.Namespace), client.MatchingLabels{"owner": cr.Name}); err != nil {
+		return nil, err
+	}
+
+	// Sort the ControllerRevisions by Revision
+	if sorted {
+		sort.Slice(controllerRevisions.Items, func(i, j int) bool {
+			return controllerRevisions.Items[i].Revision < controllerRevisions.Items[j].Revision
+		})
+	}
+
+	return controllerRevisions, nil
+}
+
+func (r *CustomReplicaSetReconciler) manageControllerRevisionHistory(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet) error {
+	jsonSpec, err := convertCRSpecToJson(cr.Spec)
+	if err != nil {
+		return err
+	}
+
+	newRevisionData := runtime.RawExtension{Raw: jsonSpec}
+	newRevisionHash, err := hashControllerRevisionData(newRevisionData)
+	if err != nil {
+		return err
+	}
```

Comment:
You need to check the hash against the one saved in cr label to avoid unnecessary full scan of all controller revisions.

Reply:
I had this functionality initially but found edge case where if someone deletes the controllerrevision the cr label is referring to a controller revision object that doesn't exist anymore. Because cr label is being maintained by us manually.

Reply:
In case you cannot find it, you just treat it as hash mismatch.

Reply:
Ok added in latest commit!